javascript/ql/src/semmle/javascript/security/dataflow/RemoteFlowSources.qll

```diff
@@ -10,8 +10,19 @@ import semmle.javascript.security.dataflow.DOM
 abstract class RemoteFlowSource extends DataFlow::Node {
   /** Gets a string that describes the type of this remote flow source. */
   abstract string getSourceType();
-}
 
+  /**
+   * Holds if this flow source comes from an incoming request, and this part of the
```

Comment:
This sounds too special-case to put it on a fairly general class like `RemoteFlowSource`. Is there a more general way to describe this scenario independent of the specifics of HTTP requests? Or should we perhaps only have this predicate on `RequestInputAccess` and specialise the sources of ReflectedXSS and ServerSideUrlRedirect to be request input accesses? (I don't think any of the other remote flow sources are relevant for those two queries, but it's of course worth checking.)

Reply:
Originally I wanted to describe it more generally as something like:

> Holds if data from this source may be controlled by an actor other than the one who sent it.

However, now we have to decide how to classify browser-side sources like `document.location`. Who sent this data? By what definition is this even a "remote" flow source?

On one hand, one could argue that they were "sent" to the user by another web server from where the user was redirected (most CSRF attacks essentially start that way). From that viewpoint, they are indeed controlled by the sender of the data (the malicious web site), thus is not third-party controlled.

On the other hand, one could argue that any flow source of interest on the browser-side must come from a third party because the client trusts both himself and the server. Thus one would expect to use `isThirdPartyControllable()` in DomBasedXss and ClientSideUrlRedirect.

In any case, it was sufficiently non-obvious how to classify `document.location`, and that's why I opted for a definition that's less ambiguous.

Reply:
I like you suggestion on moving it into `RequestInputAccess`, as that's essentially the same as saying it must be an "incoming request" (albeit restricted to HTTP requests).